release-20.1: cmd: wholesale replace roachprod, roachtest, workload from master

adityamaru opened the pull request:
Informs: #51897

adityamaru committed:
cmd: replacing release-20.1 version of roachprod, roachtest, workload(ccl)

This change does the following:

- Deletes release-20.1 roachprod, roachtest, workload and workloadccl
  packages.

- Adds the above packages from master.

- Make all the above packages compatible with 20.1. I have highlighted
  some of the major changes in each package and will inline links to why
  they were changed on master. I took some decisions on what required a
  backport and what looked too invasive, but am open to suggestions
  here.

roachtest:
- Deleted roachtests with min version 20.2.0 namely:
	- backup/KMS
	- join-init/mixed
	- disk_full
	- multitenant acceptance
	- tpchvec/bench
	- tpchvec/perf_no_stats
- Backported `tpcdsvec` roachtest

workload:
- Backported `schemachange` workload
- deleted geospatial.go
- switched all `rowenc` pkg references back to `sqlbase`
- type representation in 20.2 differed from what was supported in 20.1.
  All workloads had to be switched back to using `colTypes`

workloadccl:
- More type representation related changes
- Set --deprecated-fk-indexes to be always true for tpcc workload

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55892)
<!-- Reviewable:end -->

adityamaru commented:
Two incompatible roachtests deleted:
`backup/KMS` - which I can sign off for
`tpcdsvec` - which Yahor will sign off for

`schemachange` workload was deleted too as this had several 20.2 dependancies.

adityamaru commented on pkg/cmd/roachtest/mixed_version_join_init.go:
```diff
@@ -166,9 +165,9 @@ func unsuccessfullyAddNodeStep(nodes nodeListOption, joinNode int, newVersion st
-			if !errors.Is(err, server.ErrIncompatibleBinaryVersion) {
-				t.Fatalf("expected err: %s, got %v", server.ErrIncompatibleBinaryVersion, err)
-			}
+			// TODO(adityamaru): Check whether we want to backport this particular
```
@irfansharif this test wasn't backported to 20.1. The only incompatibility seems to be `ErrIncompatibleBinaryVersion` not being an error mode in 20.1. Just wanted to run this by you, whether this is fine as is?

irfansharif replied:
Skipping this test in 20.1 makes sense. It only makes sense for v20.2. It looks like the min version would bar this from running already. Is that not what you saw happen?

```
	r.Add(testSpec{
		Name:       "join-init/mixed",
		Owner:      OwnerKV,
		MinVersion: "v20.2.0",
```

adityamaru replied:
ahh sorry I didnt catch that. I'm just going to delete it in that case if that's fine?

irfansharif replied:
yea, that's more than fine.

adityamaru commented:
@yuzefovich the last commit is RFAL. I ran it and it passes.

yuzefovich reviewed (commented):
Thanks, LGTM.

> Reviewed 1 of 1 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55892#-:-MKLZUKq2djVamvDP15J:bnpyfmg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @jlinder)

___
*[pkg/cmd/roachtest/tpcdsvec.go, line 149 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/55892#-MKLZBG839h8-HfcvytM:-MKLZBG839h8-HfcvytN:b-qql0be) ([raw file](https://github.com/cockroachdb/cockroach/blob/1e6d4d4b43ca2f46365715f7b2466b7a074fdbbe/pkg/cmd/roachtest/tpcdsvec.go#L149)):*
> ```Go
> 			// emit an error because of that difference.
> 			if err := cmpconn.CompareConns(
> 				ctx, timeout, conns, "", "SHOW vectorize;"); err == nil {
> ```

super nit: it's an unusual break of this condition into two lines - I'd either combine this and the previous line into one or break this line into two.


<!-- Sent from Reviewable.io -->

adityamaru commented on pkg/testutils/skip/stress.go:
```diff
@@ -1,21 +0,0 @@
-// Copyright 2020 The Cockroach Authors.
```
deleting this file which was backported in  #55905, because we already have a file in `testutils/stress.go` which does this. Switching all unit tests to use this method is not in the scope of this PR and has already been done for 20.2+

adityamaru commented:
sorry for the force push noise, I'll ping the PR once CI behaves.

adityamaru commented:
I think this is ready for review, I can't figure out why the unit test is failing remotely but passing locally, and why lint thinks my workspace isn't clean. I'll work on getting CI green but would love to get some 👀 on the commits.

adityamaru commented:
@ajwerner adding you for the last commit, which backports the schemachange workload.

irfansharif commented:
> why lint thinks my workspace isn't clean.

```diff
diff --git a/pkg/workload/schemachange/optype_string.go b/pkg/workload/schemachange/optype_string.go
[14:00:19]
index e486a1b5ed..95858aa212 100644
[14:00:19]
--- a/pkg/workload/schemachange/optype_string.go
[14:00:19]
+++ b/pkg/workload/schemachange/optype_string.go
[14:00:19]
@@ -37,9 +37,9 @@ func _() {
[14:00:19]
 	_ = x[validate-26]
[14:00:19]
 }
[14:00:19]
 
[14:00:19]
-const _opType_name = "addColumnaddConstraintcreateIndexcreateSequencecreateTablecreateTableAscreateViewcreateEnumcreateSchemadropColumndropColumnDefaultdropColumnNotNulldropColumnStoreddropConstraintdropIndexdropSequencedropTabledropViewdropSchemarenameColumnrenameIndexrenameSequencerenameTablerenameViewsetColumnDefaultsetColumnNotNullsetColumnTypeinsertRowvalidate"
[14:00:19]
+const _opType_name = "addColumnaddConstraintcreateIndexcreateSequencecreateTablecreateTableAscreateViewdropColumndropColumnDefaultdropColumnNotNulldropColumnStoreddropConstraintdropIndexdropSequencedropTabledropViewdropSchemarenameColumnrenameIndexrenameSequencerenameTablerenameViewsetColumnDefaultsetColumnNotNullsetColumnTypeinsertRowvalidate"
[14:00:19]
 
[14:00:19]
-var _opType_index = [...]uint16{0, 9, 22, 33, 47, 58, 71, 81, 91, 103, 113, 130, 147, 163, 177, 186, 198, 207, 215, 225, 237, 248, 262, 273, 283, 299, 315, 328, 337, 345}
[14:00:19]
+var _opType_index = [...]uint16{0, 9, 22, 33, 47, 58, 71, 81, 91, 108, 125, 141, 155, 164, 176, 185, 193, 203, 215, 226, 240, 251, 261, 277, 293, 306, 315, 323}
[14:00:19]
 ```

Looks like it's finding the following diff. Did you run `make generate` at after having changed out the `workload` wholesale?

irfansharif commented on pkg/cmd/roachtest/backup.go:
```diff
@@ -82,120 +79,6 @@ func registerBackup(r *testRegistry) {
-		Name:       fmt.Sprintf("backup/KMS/%s", KMSSpec.String()),
-		Owner:      OwnerBulkIO,
-		Cluster:    KMSSpec,
-		MinVersion: "v20.2.0",
```
I assume you've done this already but we basically want to remove _all_ files that had a minimum version less as 20.2 (and above).

adityamaru replied:
ack, I had missed a few going to list all the files and roachtests I have deleted in the commit message.

irfansharif commented on pkg/cmd/roachtest/bank.go:
```diff
@@ -327,6 +327,29 @@ func (s *bankState) startSplitMonkey(ctx context.Context, d time.Duration, c *cl
 	}()
 }
 
+// IsExpectedRelocateError maintains an allowlist of errors related to
```
Where did these diffs come from? Do you mind adding references to them in the commit message?

irfansharif replied:
Works for me!

irfansharif commented on pkg/workload/bench_test.go:
```diff
@@ -12,10 +12,10 @@ package workload_test
 
 import (
 	"fmt"
+	"github.com/cockroachdb/cockroach/pkg/col/coltypes"
```
[nit] Imports will need to move down.

irfansharif reviewed (commented):
Thanks for taking all of this on, this looks like it was a ton of work to get just right. One thing I'm having difficulty in reviewing is actually making sense of the diffs themselves. Do you mind outlining in text for me exactly all the things we did in this PR? I think I would understand it all best if:
- it was one squashed commit (which we'll want anyway to make sure none of the intermediate commits are broken, for bisectability)
- for each diff between release-20.1 code and this new wholesale imported code, we had links to point to to understand why it was needed. If a test was written incorrectly, and we had fixed it in master, I would point to the fix PR itself. If we refactored some test code (like adding better github project routing for e.g.), I'd point to that PR (if we're deciding to keep it, which looks like we didn't?). That sort of thing.

I also don't mean to make it any more tedious for you than it already is. I just remember from past experiences that small changes to our roachtest infrastructure can start lots of little fires throughout (Ref: https://github.com/cockroachdb/jepsen/pull/24). This is a very big change, so we'd want to be very careful.

adityamaru commented:
> Thanks for taking all of this on, this looks like it was a ton of work to get just right. One thing I'm having difficulty in reviewing is actually making sense of the diffs themselves. Do you mind outlining in text for me exactly all the things we did in this PR? I think I would understand it all best if:
> 
> * it was one squashed commit (which we'll want anyway to make sure none of the intermediate commits are broken, for bisectability)
> * for each diff between release-20.1 code and this new wholesale imported code, we had links to point to to understand why it was needed. If a test was written incorrectly, and we had fixed it in master, I would point to the fix PR itself. If we refactored some test code (like adding better github project routing for e.g.), I'd point to that PR (if we're deciding to keep it, which looks like we didn't?). That sort of thing.
> 
> I also don't mean to make it any more tedious for you than it already is. I just remember from past experiences that small changes to our roachtest infrastructure can start lots of little fires throughout (Ref: [cockroachdb/jepsen#24](https://github.com/cockroachdb/jepsen/pull/24)). This is a very big change, so we'd want to be very careful.

That makes a lot of sense, I've squashed all the commits and fleshed out the commit message to give a high level overview of the major changes. Going to go in and link the diffs from master inline.

adityamaru commented on pkg/cli/cli.go:
```diff
@@ -28,12 +28,13 @@ import (
+	_ "github.com/cockroachdb/cockroach/pkg/workload/examples"     // registers workloads
+	_ "github.com/cockroachdb/cockroach/pkg/workload/kv"           // registers workloads
+	_ "github.com/cockroachdb/cockroach/pkg/workload/movr"         // registers workloads
+	_ "github.com/cockroachdb/cockroach/pkg/workload/schemachange" // registers workloads
```
calling out that schemachange needed to be registered as a workload.

adityamaru commented on pkg/cmd/roachtest/activerecord.go:
```diff
@@ -0,0 +1,237 @@
+// Copyright 2020 The Cockroach Authors.
```
https://github.com/cockroachdb/cockroach/pull/51429

adityamaru commented on pkg/cmd/roachtest/allocator.go:
```diff
@@ -15,11 +15,10 @@ import (
 	gosql "database/sql"
```
https://github.com/cockroachdb/cockroach/pull/54688

adityamaru commented on pkg/cmd/roachtest/alterpk.go:
```diff
@@ -151,9 +150,14 @@ func registerAlterPK(r *testRegistry) {
 		m.Wait()
 
 		// Run the verification checks of the TPCC workload post primary key change.
+		expensiveChecksArg := ""
```
https://github.com/cockroachdb/cockroach/pull/48067

adityamaru commented on pkg/cmd/roachtest/alterpk.go:
```diff
@@ -170,12 +174,25 @@ func registerAlterPK(r *testRegistry) {
 	})
 	r.Add(testSpec{
-		Name:  "alterpk-tpcc",
+		Name:  "alterpk-tpcc-250",
```
two tests below:
https://github.com/cockroachdb/cockroach/pull/43935

adityamaru commented on pkg/cmd/roachtest/autoupgrade.go:
```diff
@@ -0,0 +1,268 @@
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
```
https://github.com/cockroachdb/cockroach/pull/47065

adityamaru commented on pkg/cmd/roachtest/backup.go:
```diff
@@ -18,38 +18,56 @@ import (
 )
 
 func registerBackup(r *testRegistry) {
+	importBankData := func(ctx context.Context, rows int, t *test, c *cluster) string {
```
This was just a refactor which I had done, there was no logic change - https://github.com/cockroachdb/cockroach/pull/52893

adityamaru commented on pkg/cmd/roachtest/cdc.go:
```diff
@@ -505,96 +481,74 @@ func runCDCSchemaRegistry(ctx context.Context, t *test, c *cluster) {
-		Owner:      `cdc`,
-		MinVersion: "v2.1.0",
-		Cluster:    makeClusterSpec(4, cpu(16)),
+		Name:    fmt.Sprintf("cdc/tpcc-1000"),
```
https://github.com/cockroachdb/cockroach/pull/52928

adityamaru commented on pkg/cmd/roachtest/clearrange.go:
```diff
@@ -28,7 +29,7 @@ func registerClearRange(r *testRegistry) {
 			Timeout:    5*time.Hour + 90*time.Minute,
 			MinVersion: "v19.1.0",
-			Cluster:    makeClusterSpec(10),
+			Cluster:    makeClusterSpec(10, cpu(16)),
```
https://github.com/cockroachdb/cockroach/pull/53572

adityamaru commented on pkg/cmd/roachtest/cluster.go:
```diff
@@ -52,18 +53,21 @@ const (
-	clusterWipe  bool
-	zonesF       string
-	teamCity     bool
+	local            bool
```
there is a whole bunch of work that has gone on in this file - https://github.com/cockroachdb/cockroach/blame/master/pkg/cmd/roachtest/cluster.go

not sure how to provide any isolated context to the changes.

adityamaru commented on pkg/cmd/roachtest/cluster_test.go:
```diff
@@ -66,12 +66,62 @@ func (t testWrapper) logger() *logger {
 	return nil
 }
 
+func TestExecCmd(t *testing.T) {
```
https://github.com/cockroachdb/cockroach/pull/49371

adityamaru commented on pkg/cmd/roachtest/django.go:
```diff
@@ -16,9 +16,11 @@ import (
 )
 
-var djangoReleaseTagRegex = regexp.MustCompile(`^(?P<major>\d+)\.(?P<minor>\d+)\.(?P<point>\d+)$`)
+var djangoReleaseTagRegex = regexp.MustCompile(`^(?P<major>\d+)\.(?P<minor>\d+)(\.(?P<point>\d+))?$`)
```
This test was added in - https://github.com/cockroachdb/cockroach/pull/41516

adityamaru commented on pkg/cmd/roachtest/drop.go:
```diff
@@ -160,9 +160,7 @@ func registerDrop(r *testRegistry) {
-
-	// 1GB
-	initDiskSpace := int(1e9)
+	initDiskSpace := 256 << 20 // 256 MB
```
https://github.com/cockroachdb/cockroach/pull/55063

adityamaru commented on pkg/cmd/roachtest/follower_reads.go:
```diff
@@ -112,7 +112,7 @@ func runFollowerReadsTest(ctx context.Context, t *test, c *cluster) {
 		panic("data is empty")
 	}
-	verifySelect := func(ctx context.Context, node, k int, expectError bool, expectedVal int64) func() error {
+	verifySelect := func(ctx context.Context, node, k int, expectedVal int64) func() error {
```
Flake fixes - https://github.com/cockroachdb/cockroach/pull/54622

adityamaru commented on pkg/cmd/roachtest/import.go:
```diff
@@ -62,12 +61,12 @@ func registerImportTPCC(r *testRegistry) {
 	const geoWarehouses = 4000
 	const geoZones = "europe-west2-b,europe-west4-b,asia-northeast1-b,us-west1-b"
 	r.Add(testSpec{
+		Skip:    "#37349 - OOMing",
```
I'll bring this up with bulk, with Pebble this can probably be unskipped.

adityamaru replied:
again, 20.1 doesn't use Pebble I got confused. Nothing to do on this branch.

adityamaru commented on pkg/cmd/roachtest/inconsistency.go:
```diff
@@ -22,6 +22,7 @@ func registerInconsistency(r *testRegistry) {
 	r.Add(testSpec{
 		Name:       fmt.Sprintf("inconsistency"),
 		Owner:      OwnerKV,
+		Skip:       "Uses RocksDB put command; unskip when that's bypassed",
```
TODO(adityamaru): https://github.com/cockroachdb/cockroach/commit/8732d4281a7f913c53a5a069932d9caa222dcd57#diff-d0816e2d21e74d24bc9c81ef3e8b14ebf0871e73bdb4045dfbbf6e4cea69f920 was merged after this PR picked its roachtest.

adityamaru replied:
oh nevermind, 20.1 still uses rocksdb. Nothing to do here.

adityamaru commented on pkg/cmd/roachtest/kv.go:
```diff
@@ -254,9 +247,9 @@ func registerKVContention(r *testRegistry) {
-				// dipped below 100 qps for 5% of the time.
-				const minQPS = 100
-				verifyTxnPerSecond(ctx, c, t, c.Node(1), start, end, minQPS, 0.05)
+				// dipped below 50 qps for 10% of the time.
```
#54230

adityamaru commented on pkg/cmd/roachtest/many_splits.go:
```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 The Cockroach Authors.
```
#47143

adityamaru commented on pkg/cmd/roachtest/mixed_version_decommission.go:
```diff
@@ -0,0 +1,293 @@
+// Copyright 2020 The Cockroach Authors.
```
#50329

adityamaru commented on pkg/cmd/roachtest/mixed_version_schemachange.go:
```diff
@@ -0,0 +1,145 @@
+// Copyright 2020 The Cockroach Authors.
```
#47362

adityamaru commented on pkg/cmd/roachtest/mixed_version_jobs.go:
```diff
@@ -0,0 +1,334 @@
+// Copyright 2020 The Cockroach Authors.
```
#47552

adityamaru commented on pkg/cmd/roachtest/namespace_upgrade.go:
```diff
@@ -0,0 +1,277 @@
+// Copyright 2020 The Cockroach Authors.
```
#49099

adityamaru commented on pkg/cmd/roachtest/orm_helpers.go:
```diff
@@ -32,54 +32,64 @@ func alterZoneConfigAndClusterSettings(
 
 	if _, err := db.ExecContext(
-		ctx, `ALTER DATABASE system CONFIGURE ZONE USING num_replicas = 1, gc.ttlseconds = 120;`,
+		ctx, `ALTER DATABASE system CONFIGURE ZONE USING num_replicas = 1, gc.ttlseconds = 60;`,
```
#53606

adityamaru commented on pkg/cmd/roachtest/pebble.go:
```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 The Cockroach Authors.
```
#50892

adityamaru commented on pkg/cmd/roachtest/quit.go:
```diff
@@ -0,0 +1,463 @@
+// Copyright 2020 The Cockroach Authors.
```
#45149

adityamaru commented on pkg/cmd/roachtest/replicagc.go:
```diff
@@ -12,31 +12,40 @@ package main
-			runReplicaGCChangedPeers(ctx, t, c, false /* withRestart */)
-		},
-	})
+	for _, restart := range []bool{true, false} {
```
#51394

adityamaru commented on pkg/cmd/roachtest/secondary_indexes.go:
```diff
@@ -12,30 +12,79 @@ package main
+// runIndexUpgrade runs a test that creates an index before a version upgrade,
+// and modifies it in a mixed version setting. It aims to test the changes made
+// to index encodings done to allow secondary indexes to respect column families.
+func runIndexUpgrade(ctx context.Context, t *test, c *cluster, predecessorVersion string) {
```
#49878

adityamaru commented on pkg/cmd/roachtest/sqlalchemy.go:
```diff
@@ -19,216 +19,196 @@ import (
-		if err != nil {
-			t.Fatal(err)
-		}
+	r.Add(testSpec{
```
#51558

adityamaru commented on pkg/workload/schemachange/optype_string.go:
```diff
@@ -0,0 +1,49 @@
+// Code generated by "stringer -type=opType"; DO NOT EDIT.
```
@ajwerne I had to drop enum related logic from the schemachange workload to make it 20.1 compatible. The squash hides the compatibility changes I had to make - would you prefer this to be an independent PR where I add masters version of schemachange, and then make the compatibility tweaks?

irfansharif reviewed (approved):
I think the changes here LGTM, assuming it all checks out in CI (I'm unsure how to make sure we're running all roachtests in CI for release-20.1, I assume that's what you've been doing?). Major plause for taking all of this one, this must have been a slog. Linking in the referencing PRs really helped for the review. Hopefully it helped isolate the points where we needed more manual intervention, in case we need to revisit going forward.


I'm glad most of it were clean(-ish) backports/wholesale replacement. I think that'll make future backports to roachtest/roachprod easy enough to not want to run them off of master again. I'll stamp off here, and leave you to discuss with Andrew regarding what's to be done with the schamachange workload diffs.

jlinder commented:
> I think the changes here LGTM, assuming it all checks out in CI (I'm unsure how to make sure we're running all roachtests in CI for release-20.1, I assume that's what you've been doing?). 

The changes that need to happen in CI:

- Change the script in the [Roachtest Nightly GCE](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_WorkloadNightly?buildTypeTab=buildTypeSettings#all-projects) build config to use the branch it was triggered for (and not master)
- Change the script in the [Roachtest Nightly AWS](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyAws?buildTypeTab=buildTypeSettings) build config to use the branch it was triggered for (and not master)

These changes need to happen in concert with the [19.2](https://github.com/cockroachdb/cockroach/pull/55865) and the [20.2](https://github.com/cockroachdb/cockroach/pull/55899) PRs. 

The roachtest build config already runs the roachtests from its branch and does not need to be changed.

adityamaru commented:
Added `roachprod-stress` and the two nightly build scripts to this PR too so that we can avoid having to reference master completely. Also rebased it on top of the lint util backport - #56123.   

I'm not sure how to deal with the flaking `TestAllRegisteredImportFixture/tpcc`, its been an evasive flake which noone from bulk has been able to reproduce as can be seen in #47015, and it hasn't flaked in 2 months. I ran it on roachprod-stressrace for 35 mins with no failure.

ajwerner reviewed (commented):
Schema change workload <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 34 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55892#-:-MKuoaeX345k-EluopFe:b-q5fein)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @jlinder, and @yuzefovich)



<!-- Sent from Reviewable.io -->

adityamaru commented:
CI is green woop!

I'm going to merge this, and then make a switch to the build template to special case `release-20.1` to run on itself. cc: @jlinder

adityamaru commented:
Ran a special-cased run on TC for this branch to check that it passes a few GCE nightlies. Going to merge and trigger a nightly GCE run on release-20.1.
